File: benchmarks/fixtures/webapp_rb/services/cacheable.rb
# Cacheable module for services with built-in caching.

require_relative '../utils/helpers'

# Mixin providing cache get/set/invalidate to any service.
module Cacheable
  def cache_get(key)
    @cache_store ||= {}
    entry = @cache_store[key]
    if entry && Time.now.to_i < entry[:expiry]
      get_logger('cacheable').info("Cache hit: #{key}")
      return entry[:value]
    end
    @cache_store.delete(key) if entry
    get_logger('cacheable').info("Cache miss: #{key}")
    nil
  end

  def cache_set(key, value, ttl = 300)
    @cache_store ||= {}
    @cache_store[key] = { value: value, expiry: Time.now.to_i + ttl }
    get_logger('cacheable').info("Cache set: #{key} (ttl=#{ttl}s)")
  end

  def cache_invalidate(key)
    @cache_store ||= {}
    !@cache_store.delete(key).nil?
  end

  def cache_clear
    @cache_store ||= {}
    count = @cache_store.size
    @cache_store.clear
    get_logger('cacheable').info("Cache cleared: #{count} entries")
    count
  end
end

File: benchmarks/fixtures/webapp_ts/src/services/cacheable.ts
/** Cacheable service with built-in caching. */
import { getLogger } from '../utils/helpers';
import { BaseService } from '../auth/service';

const logger = getLogger("services.cacheable");

/** Service with caching support. */
export class CacheableService extends BaseService {
    private cache: Map<string, { value: unknown; expiry: number }> = new Map();
    private defaultTtl: number = 300;

    constructor(serviceName: string = "cacheable") {
        super(serviceName);
    }

    /** Get value from cache. */
    cacheGet(key: string): unknown | null {
        const entry = this.cache.get(key);
        if (entry && Date.now() < entry.expiry) {
            logger.info(`Cache hit: ${key}`);
            return entry.value;
        }
        if (entry) this.cache.delete(key);
        logger.info(`Cache miss: ${key}`);
        return null;
    }

    /** Set value in cache. */
    cacheSet(key: string, value: unknown, ttl?: number): void {
        const effectiveTtl = ttl ?? this.defaultTtl;
        this.cache.set(key, { value, expiry: Date.now() + effectiveTtl * 1000 });
        logger.info(`Cache set: ${key} (ttl=${effectiveTtl}s)`);
    }

    /** Invalidate a cache entry. */
    cacheInvalidate(key: string): boolean {
        return this.cache.delete(key);
    }

    /** Clear all cache entries. */
    cacheClear(): number {
        const count = this.cache.size;
        this.cache.clear();
        logger.info(`Cache cleared: ${count} entries`);
        return count;
    }
}
